spec/index.bs

```diff
@@ -726,8 +720,62 @@ interface GPUBufferUsage {
 };
 </script>
 
+There are no restrictions on combining buffer usage flags, but some combinations
+man incur additional overhead. (For example, MAP_READ and STORAGE may incur a
+higher than expected overhead of copies, as the implemention must keep MAP_READ
+buffer mapping data as fresh as Fences require)
+
 ## Buffer Mapping ## {#buffer-mapping}
 
+<script type=idl>
+typedef unsigned long GPUMapFlags;
+interface GPUMap {
+    const GPUMapFlags READ  = 0x0001;
+    const GPUMapFlags WRITE = 0x0002;
+};
+</script>
+
+## {{GPUBuffer/map()}} ##
+
+If the devices is lost, returns null.
+
+If {{GPUBuffer/[[state]]}} is "executing", if {{GPUBuffer/[[readyAfterFenceID]]}}
+is greater than {{GPUQueue/[[readyFenceID]]}}, generate an error and return null.
+
+Otherwise, {{GPUBuffer/[[state]]}} becomes "ready".
+
+If {{GPUBuffer/[[state]]}} is not "ready", returns null.
+(Only one mapping may be active per buffer at a time)
+
+If {{GPUBuffer/map()}} includes {{GPUMap/READ}} but the buffer is not marked for
+{{GPUBufferUsage/MAP_READ}}, generate an error and return null.
+If {{GPUBuffer/map()}} includes {{GPUMap/WRITE}} but the buffer is not marked for
+{{GPUBufferUsage/MAP_WRITE}}, generate an error and return null.
+
+When buffers are used in a {{GPUCommandBuffer}} sent successfully to
+{{GPUQueue.submit()}}, {{[[state]]}} becomes "executing", and {{[[readyAfterFenceID]]}}
+becomes {{GPUQueue/[[nextFenceID]]}}.
+
+On {{GPUQueue/createFence()}} and {{GPUQueue/signal()}}:
+* In {{GPUFence/[[pendingfenceIDBySignalValue]]}}, assign key signalValue to value
+  {{GPUQueue/[[nextFenceID]]}}.
+* Increment {{GPUQueue/[[nextFenceID]]}} by 1.
+
+When the Promise from {{GPUFence.onCompletion()}} resolves:
+* Set {{GPUQueue/[[readyFenceID]]}} to the larger of {{GPUQueue/[[readyFenceID]]}} or the value
+  of {{GPUFence/[[pendingfenceIDBySignalValue]]}} at key signalValue.
+* Remove {{GPUFence/[[pendingfenceIDBySignalValue]]}} at key signalValue.
+
+If {{GPUBuffer/map()}} does not include {{GPUMap/READ}}, reads from the ArrayBuffer are zeros.
```

Comment:
It's not mentioned anywhere but I assume that if the flags include `GPUMap_READ` then the content of the `ArrayBuffer` is the content of the `[[buffer]]` at that point.

There is no signal from the application whether it wants to read the content of the buffer or wait for later. And to guarantee the semantics outlined in this PR, the browser would have to eagerly send the content of each `ready` `MAP_READ` buffer, after each fence is passed so that when the user sees the fence passed, the content of the buffer is already on the content process side.

Reply:
I agree with this, even though in many cases reading the buffer when it's "ready" is totally fine:
  - either the user keeps using the buffer on GPU and it never actually reaches "ready"
  - or the user really needs to read from it, hence the copy is needed

I wonder if this particular issue can be addressed by making the "request to read" explicit in the API as a queue operation, i.e. the user says `queue.makeBufferReadable(buffer, range)`, and if they put a fence after that they have a guarantee that the buffer can be read.

Have to acknowledge that this looks very similar to a hypothetical `queue.readBufferData` that we could to #509 in line with the API it exposes.

Reply:
Yeah, basically we need some form of explicit asynchrony and any solution here will be 
conceptually similar to `mapReadAsync` or `queue.readBufferData`.

Reply:
`queue.readBufferData` could work one of two ways:

1. It returns a `Promise<ArrayBuffer>`, or
2. It doesn't return anything, and after the app waits on a fence, they can synchronously get the results.

If we're going to do https://github.com/gpuweb/gpuweb/pull/509/files `queue.writeBuffer()` and `queue.writeTexture()`, it makes sense to add `queue.readBufferData()` for symmetry.

Reply:
Just a wild idea if we can detach an `ArrayBuffer` passed to `readBufferData` and then "re-attach" it whenever the promise resolves and it has the data we've just written. That would allow the user to avoid creating new objects.